src/System.IO.Pipelines.Networking.Sockets/Internal/Signal.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System.IO.Pipelines.Networking.Sockets.Internal
+{
+    /// <summary>
+    /// Very lightweight awaitable gate - intended for use in high-volume single-producer/single-consumer
+    /// scenario, in particular targeting the bridge between async IO operations
+    /// and the async method that is pumping the read/write queue. A key consideration is that
+    /// no objects (in particular Task/TaskCompletionSource) are allocated even in the await case. Instead,
+    /// a custom awaiter is provided. Works like a <see cref="ManualResetEvent "/> - the <see cref="Reset"/> method must
+    /// be called between operations.
+    /// </summary>
+    internal class Signal : INotifyCompletion
+    {
+        private readonly ContinuationMode _continuationMode;
+
+        private Action _continuation;
+        private static readonly Action _completedSentinel = delegate { };
+
+        public Signal(ContinuationMode continuationMode = ContinuationMode.Synchronous)
+        {
+            _continuationMode = continuationMode;
+        }
+
+        public bool IsCompleted => ReferenceEquals(_completedSentinel, Volatile.Read(ref _continuation));
+
+        private object SyncLock => this;
+
+        public Signal GetAwaiter() => this;
+
+        public void GetResult() { }
+
+        public void OnCompleted(Action continuation)
+        {
+            if (continuation != null)
```

Comment:
> if (continuation != null) [](start = 11, length = 26)

This would only happen if it was misused... the compiler will never pass a null continuation.